src/lib/crypto/kimchi_backend/gadgets/and.ml

```diff
@@ -0,0 +1,148 @@
+open Core_kernel
+
+open Kimchi_backend_common.Plonk_constraint_system.Plonk_constraint
+
+(* AND *)
+
+(* Boolean And of length bits *)
+let band (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (input1 : Circuit.Field.t) (input2 : Circuit.Field.t) (length : int)
+    (len_xor : int) : Circuit.Field.t =
+  let open Circuit in
+  let output_and =
+    exists Field.typ ~compute:(fun () ->
+        (* Recursively build And gadget with leading Xors and a final Generic gate *)
+        (* It will also check the right lengths of the inputs, no need to do it again *)
+        let xor_output =
+          Xor.bxor (module Circuit) input1 input2 length len_xor
+        in
+
+        (* Read inputs *)
+        let input1_field =
+          Common.cvar_field_to_field_as_prover (module Circuit) input1
+        in
+        let input2_field =
+          Common.cvar_field_to_field_as_prover (module Circuit) input2
+        in
+
+        (* Convert inputs field elements to list of bits of length the field size *)
+        let input1_bits = Field.Constant.unpack @@ input1_field in
+        let input2_bits = Field.Constant.unpack @@ input2_field in
+
+        (* AND list of bits to obtain output *)
+        let and_bits =
+          List.map2_exn input1_bits input2_bits ~f:(fun b1 b2 -> b1 && b2)
+        in
+
+        (* Convert back to field a AND b *)
+        let and_output = Field.Constant.project and_bits in
+
+        (* Compute sum of a + b *)
+        let sum = Field.add input1 input2 in
+
+        with_label "sum_of_inputs" (fun () ->
+            assert_
+              { annotation = Some __LOC__
+              ; basic =
+                  Kimchi_backend_common.Plonk_constraint_system.Plonk_constraint
+                  .T
+                    (Basic
+                       { l = (Field.Constant.one, input1)
+                       ; r = (Field.Constant.one, input2)
+                       ; o =
+                           ( Option.value_exn Field.(to_constant (negate one))
+                           , sum )
+                       ; m = Field.Constant.zero
+                       ; c = Field.Constant.zero
+                       } )
+              } ) ;
+
+        (* Compute AND as 2 * and = sum - xor *)
+        with_label "and_equation" (fun () ->
```

Comment:
Same comment above.  I think we should actually use our `Generic.add` gadget to make code must clearer

Reply:
but generic add does not accept outputs, only the inputs, and i want to explicitly create it with 2 * and as output

Reply:
Good point.